README.md
# [morottaja][]

Kutsuntatapa: (vaihda haluttu päivä, ja luo linkki morottaja-executableen. pnrTemplate on rosoinen purkka.)
echo -n "2020-05-18" | ./morottaja; cat pnrTemplate | ./assemblePDF.sh


CHANGELOG.md
# Change log

morottaja uses [Semantic Versioning][].
The change log is available through the [releases on GitHub][].

[Semantic Versioning]: http://semver.org/spec/v2.0.0.html
[releases on GitHub]: https://github.com/githubuser/morottaja/releases


library/Login.hs
{-# LANGUAGE DuplicateRecordFields, OverloadedStrings, RecordWildCards #-}

module Login where

import qualified Data.ByteString.Char8 as B
import qualified Data.Text as T
import Data.Time
import Data.Time.Format.ISO8601
import Network.HTTP.Conduit
import Network.HTTP.Simple
import Network.HTTP.Types

import Types

getLoginData :: LoginRequest -> IO (Maybe LoginData)
getLoginData loginReq =
  let req = setRequestMethod "POST"
        <$> setRequestBodyJSON loginReq
        <$> parseRequest "https://www.aamulehti.fi/aas/login?publication=aamulehti" :: Maybe Request
  in case req of
       (Just r) -> do
         print r
         getResponseBody <$> httpJSON r
       Nothing -> return Nothing

getMags :: MagsRequest -> IO (Maybe Mags)
getMags magsReq = do
  let magsJsonReq = parseRequest
                    ("https://www.aamulehti.fi/aas/mags?include_linked=1&publication=aamulehti&start_date="
                     ++ iso8601Show (fillInStartDate (endDate magsReq))
                     ++ "&end_date=" ++ iso8601Show (endDate magsReq))
  case magsJsonReq of
    (Just r) -> do
      getResponseBody <$> httpJSON r
      --mags <- httpJSON r :: IO (Response Mags)
      --return (Just $ getResponseBody mags)
    Nothing -> do
      putStrLn "Virheellisesti muodostettu mags.json request."
      return Nothing
  where
    fillInStartDate endDate = addUTCTime (negate $ 10 * nominalDay) endDate

getMH5 :: AamulehtiMag -> LoginData -> IO (Maybe MH5Cookies)
getMH5 mag lData =
  let magUuid = T.unpack $ uuid mag
      user = T.unpack $ username (lData :: LoginData)
      req :: Maybe Request
      req = addAuthHeader (B.pack $ T.unpack $ token lData)
        <$> parseRequest ("https://www.aamulehti.fi/aas/mags/" ++ magUuid
                         ++ "?publication=aamulehti&username=" ++ user)
  in case req of
       (Just r) -> do
         redirectUrl <- getResponseBody <$> httpJSON r
         let mh5Req = parseRequest (T.unpack $ redirect redirectUrl)
         case mh5Req of  -- TODO: tarkastele löytyykö sekä mh5_tok että mh5_ret
           (Just mh5R) -> do
             gotCookies <- responseCookieJar <$> httpBS mh5R
             return $ Just gotCookies
           Nothing -> do
             putStrLn "Richiefin redirect URLin kanssa ongelma."
             return Nothing
       Nothing -> do
         putStrLn "Virheellisesti muodostettu MH5 request."
         return Nothing
  where
    addAuthHeader :: B.ByteString -> Request -> Request
    addAuthHeader token = addRequestHeader hAuthorization (B.concat ["Bearer ", token])

getMagazineJson :: AamulehtiMag -> MH5Cookies -> IO (Maybe Magazine)
getMagazineJson mag mh5 = do
  let magUrl = "https://aamulehti.ap.richiefi.net" ++
               foldl (\acc x -> if cookie_name x == "mh5_tok" then B.unpack (cookie_path x) else acc) "" (destroyCookieJar mh5) ++
               "/magazine.json"
  protoReq <- parseRequest magUrl
  let req = protoReq { cookieJar = Just mh5 }
  getResponseBody <$> httpJSON req


library/Print.hs
{-# LANGUAGE DuplicateRecordFields, OverloadedStrings, RecordWildCards #-}

module Print where

import Control.Concurrent (threadDelay) -- delay in microseconds
import qualified Data.ByteString as B
import Data.Maybe (catMaybes, fromJust, isJust)
import qualified Data.Text as T
import Network.HTTP.Simple
import System.Random (randomRIO)

import Types

getMagazine :: Magazine -> IO CompleteAamulehti
getMagazine Magazine{..} = do
  let numberedSpreads = zip [0, 2..] spreads
  completeAamulehti <- mapM (\(pnr, spr) -> getSpread pnr baseurl spr) numberedSpreads
  return $ CompleteAamulehti
    (concatMap (\(a, b) -> catMaybes [a, b]) completeAamulehti)

getSpread :: PageNumber -> T.Text -> Spread -> IO (Maybe AamulehtiPage,
                                                   Maybe AamulehtiPage)
getSpread pnr baseUrl Spread{..} = do
  leftSide <- mapM (getPage baseUrl) leftPage
  rightSide <- mapM (getPage baseUrl) rightPage
  mimicAHumanAndWaitBetweenSpreads
  return (if isJust leftSide
          then Just $ AamulehtiPage (head $ fromJust leftSide) (head $ tail $ fromJust leftSide) (pnr)
          else Nothing,
          if isJust rightSide
          then Just $ AamulehtiPage (head $ fromJust rightSide) (head $ tail $ fromJust rightSide) (pnr + 1)
          else Nothing) -- TODO: list <-> tuple wankkaus pois

getPage :: T.Text -> (T.Text, T.Text) -> IO [B.ByteString]
getPage baseUrl (page, overlay) = sequence [dlUrl (T.concat [baseUrl, "/", page]),
                                            dlUrl (T.concat [baseUrl, "/", overlay])]


dlUrl :: T.Text -> IO B.ByteString
dlUrl pageUrl = do
  req <- parseRequest $ T.unpack pageUrl
  getResponseBody <$> httpBS req

mimicAHumanAndWaitBetweenSpreads :: IO ()
mimicAHumanAndWaitBetweenSpreads = do
  die2 <- randomRIO (1, 6) :: IO Int
  die3 <- unNegateDie <$> randomRIO (-10, 10) :: IO Int
  die4 <- unNegateDie <$> randomRIO (-360, 75) :: IO Int
  let total = 1 + die2 + die3
        + die4
  putStrLn ("Sivun kääntämiseen menee " ++ show total ++ "s.")
  threadDelay (total * 1000000)
  where
    unNegateDie die = if die < 0 then 0 else die

printMagazine :: CompleteAamulehti -> IO FilePath
printMagazine CompleteAamulehti{..} = do
  mapM_ writePages pages
  -- tehdään PDF-kutsu
  return "aamulehti.pdf"
  where
    writePages :: AamulehtiPage -> IO ()
    writePages AamulehtiPage{..} =
      let pnr p
            | p < 100 && p >= 10 = "0" ++ show p
            | p < 10 = "00" ++ show p
            | otherwise = show p
      in do
        B.writeFile ("pages/page_" ++ pnr pageNumber ++ ".jpg") page
        B.writeFile ("pages/page_" ++ pnr pageNumber ++ "_overlay.png") overlay


library/Aamulehti.hs
{-# LANGUAGE DuplicateRecordFields, OverloadedStrings, RecordWildCards #-}

module Aamulehti (main) where

import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as LB
import qualified Data.Text as T
import Data.Either
import Data.Time
import Data.Time.Format.ISO8601
import System.FilePath

import Login
import Print
import Types

-- | An example function.
main :: IO ()
main = do
  --testMags <- LB.readFile jsonTestMags
  --print (eitherDecode testMags :: Either String Mags)
  --testMagazine <- LB.readFile jsonTestMagazine
  --print (eitherDecode testMagazine :: Either String Magazine)

  isoDate <- getContents

  creds <- getAamulehtiCreds
  print (encode creds)

  loginData <- getAamulehtiCreds >>= getLoginData
  print loginData

  magsEndDate <- iso8601ParseM (isoDate ++ "T20:59:59.999Z")
  let placeholderMagsCatalogReq = MagsRequest Nothing magsEndDate
  magsCatalog <- getMags placeholderMagsCatalogReq
  print magsCatalog

  case loginData of
    (Just lData) -> do
      case magsCatalog of
        (Just mc) -> do
          mh5 <- getMH5 (head $ mags mc) lData
          print mh5
          case mh5 of
            (Just mh5Cookies) -> do
              magazineJson <- getMagazineJson (head $ mags mc) mh5Cookies
              print magazineJson
              case magazineJson of
                (Just zine) ->
                  getMagazine zine >>= printMagazine >>= putStrLn
                Nothing ->
                  putStrLn "Ei saatu magazine.jsonia"
            Nothing ->
              putStrLn "MH5-keksit jäi saamatta."
        Nothing ->
          putStrLn "mags.jsonin hakeminen epäonnistui."
    Nothing ->
      putStrLn "Kirjautuminen epäonnistui."



getAamulehtiCreds :: IO LoginRequest
getAamulehtiCreds = do
  creds <- T.lines <$> T.pack <$> readFile "aamulehtiCreds"
  return $ LoginRequest (head creds) (head $ tail creds)


library/Types.hs
{-# LANGUAGE DuplicateRecordFields, OverloadedStrings, RecordWildCards #-}

module Types where

import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.Text as T
import Data.Time
import Network.HTTP.Conduit (CookieJar)

{--
Huomioi tilanteet joissa JSONin result /= 200 virhetilanteiden hienosäätöä varten; nyt riittää vain että parser palauttaa virheen.

--}
data LoginRequest =
  LoginRequest { username :: T.Text
               , password :: T.Text
               } deriving (Show)

instance ToJSON LoginRequest where
  toJSON LoginRequest{..} =
    object [
    "username" .= username,
    "password" .= password
    ]

data LoginData =
  LoginData { username :: T.Text
            , token :: T.Text
            , hash :: T.Text
            , authorized :: Bool
            } deriving (Show)

instance FromJSON LoginData where
  parseJSON = withObject "LoginData" $ \v -> do
    nestedLogin <- v .: "value"
    username <- nestedLogin .: "username"
    token <- nestedLogin .: "token"
    hash <- nestedLogin .: "hash"
    authorized <- nestedLogin .: "authorized"
    return LoginData{..}

data MagsRequest =
  MagsRequest { startDate :: Maybe UTCTime -- "2020-05-04T20:59:59.999Z"
                             -- if Nothing then endDate - 10 days mimicking the website
              , endDate :: UTCTime -- "2020-05-14T20:59:59.999Z"
              }

data AamulehtiMag = -- value > mags > Array Aamulehti
  AamulehtiMag { uuid :: T.Text
               , name :: T.Text -- == date ("14.5.2020")
               , publishedAt :: UTCTime -- "2020-05-14T00:00:00.000Z"
               } deriving (Show)

instance FromJSON AamulehtiMag where
  parseJSON = withObject "AamulehtiMag" $ \v -> do
    uuid <- v .: "uuid"
    name <- v .: "name"
    publishedAt <- v .: "publishedAt"
    return AamulehtiMag{..}

data Mags =
  Mags { mags :: [AamulehtiMag]
       } deriving (Show)

instance FromJSON Mags where
  parseJSON = withObject "Mags" $ \v -> do
    nestedFirst <- v .: "value"
    nestedSecond <- nestedFirst .: "mags"
    mags <- nestedSecond .: "Aamulehti"
    return Mags{..}

data RichiefiRedirect =
  RichiefiRedirect { redirect :: T.Text
                   }

instance FromJSON RichiefiRedirect where
  parseJSON = withObject "RichiefiRedirect" $ \v -> do
    redirect <- v .: "value"
    return RichiefiRedirect{..}

type MH5Cookies = CookieJar

data Spread =
  Spread { leftPage :: Maybe (T.Text, T.Text)
         , rightPage :: Maybe (T.Text, T.Text)
         } deriving (Show)

instance FromJSON Spread where
  parseJSON = withObject "Spread" $ \v -> do
    let maybeMap obj str = mapM (\x -> x .: str) obj -- >>= olisi elegantimpi
    nestedLeft <- v .:? "left"

    nestedLPortrait <- maybeMap nestedLeft "portrait"
    nestedLImages <- maybeMap nestedLPortrait "images"
    leftPage <- maybeMap nestedLImages "background"
    leftOverlay <- maybeMap nestedLImages "overlay"

    nestedRight <- v .:? "right"
    nestedRPortrait <- maybeMap nestedRight "portrait"
    nestedRImages <- maybeMap nestedRPortrait "images"
    rightPage <- maybeMap nestedRImages "background"
    rightOverlay <- maybeMap nestedRImages "overlay"
    return $ Spread (fuseMaybes leftPage leftOverlay) (fuseMaybes rightPage rightOverlay)
      where
        fuseMaybes (Just pg) (Just ol) = Just (pg, ol)
        fuseMaybes (Just _) Nothing = Nothing
        fuseMaybes Nothing (Just _) = Nothing
        fuseMaybes Nothing Nothing = Nothing

data Magazine = -- magazine.json > variants > html5_splitjpg_960
  Magazine { baseurl :: T.Text
           , spreads :: [Spread]
           } deriving (Show)

instance FromJSON Magazine where
  parseJSON = withObject "Magazine" $ \v -> do
    nestedFirst <- v .: "variants"
    nestedSecond <- nestedFirst .: "html5_splitjpg_960"
    baseurl <- nestedSecond .: "baseurl"
    spreads <- nestedSecond .: "spreads"
    return Magazine{..}

type PageNumber = Int

data AamulehtiPage =
  AamulehtiPage { page :: B.ByteString -- jpeg binary
                , overlay :: B.ByteString -- png binary
                , pageNumber :: PageNumber
                }

data CompleteAamulehti =
  CompleteAamulehti { pages :: [AamulehtiPage]
                    } -- TODO: date
